src/middleware/etaghasher.rs

```diff
@@ -0,0 +1,318 @@
+/// response status is `200 OK`. If you provide your own `filter`, you
+/// will want to check for these conditions as well.
+///
+/// ```rust
+/// # extern crate actix_web;
+/// use actix_web::{http, middleware, App, HttpResponse};
+/// use middleware::etaghasher::{EtagHasher, DefaultHasher, DefaultFilter};
+///
+/// fn main() {
+///     let eh = EtagHasher::new(DefaultHasher, DefaultFilter);
+///     let app = App::new()
+///         .middleware(eh)
+///         .resource("/test", |r| {
+///             r.method(http::Method::GET).f(|_| HttpResponse::Ok());
+///         })
+///         .finish();
+/// }
+/// ```
+///
+/// With custom `hash` and `filter` functions:
+///
+/// ```rust
+/// # extern crate actix_web;
+/// use actix_web::{http, middleware, App, HttpRequest, HttpResponse};
+/// use middleware::etaghasher::EtagHasher;
+///
+/// fn main() {
+///     let eh = EtagHasher::new(
+///         |_input: &[u8]| "static".to_string(),
+///         |_req: &HttpRequest<()>, _res: &HttpResponse| true,
+///     );
+///     let app = App::new()
+///         .middleware(eh)
+///         .resource("/test", |r| {
+///             r.method(http::Method::GET).f(|_| HttpResponse::Ok());
+///         })
+///         .finish();
+/// }
+/// ```
+
+use error::Result;
+use header::EntityTag;
+use httprequest::HttpRequest;
+use httpresponse::HttpResponse;
+use middleware;
+
+use std::marker::PhantomData;
+
+/// `Middleware` for generating ETag headers and returning `304 Not Modified`
+/// responses upon receipt of a matching `If-None-Match` request header.
+
+/// Can produce an ETag value from a byte slice. Per RFC 7232, **must only
+/// produce** bytes with hex values `21`, `23-7E`, or greater than or equal
+/// to `80`. Producing invalid bytes will result in a panic when the output
+/// is converted to an ETag.
+pub trait Hasher {
+    /// Produce an ETag value given a byte slice.
+    fn hash(&self, input: &[u8]) -> String;
+}
+/// Can test a (request, response) pair and return `true` or `false`
+pub trait RequestFilter<S> {
+    /// Return `true` if ETag processing should be applied to this
+    /// `(request, response)` pair and `false` otherwise. A `false` return
+    /// value will immediately return the original response unchanged.
+    fn filter(&self, req: &HttpRequest<S>, res: &HttpResponse) -> bool;
+}
+
+// Closure implementations
+impl<F: Fn(&[u8]) -> String> Hasher for F {
+    fn hash(&self, input: &[u8]) -> String {
+        self(input)
+    }
+}
+impl<S, F: Fn(&HttpRequest<S>, &HttpResponse) -> bool> RequestFilter<S> for F {
+    fn filter(&self, req: &HttpRequest<S>, res: &HttpResponse) -> bool {
+        self(req, res)
+    }
+}
+
+// Defaults
+/// Computes an ETag value from a byte slice using a default cryptographic hash
+/// function.
+pub struct DefaultHasher;
+impl Hasher for DefaultHasher {
+    fn hash(&self, input: &[u8]) -> String {
+        use sha1;
+        let mut h = sha1::Sha1::new();
```

Comment:
Wouldn't it be it better to just create `Sha1` once and store it hasher?
You'd just need to reset it on each call instead of creating always which would be cheaper

Reply:
Yes, done!